internal/generated/api/retry_http_client.go

```diff
@@ -0,0 +1,203 @@
+// Package api provides HTTP client functionality with intelligent retry logic
+// for the Couchbase Capella API. This package uses the Hashicorp retryablehttp
+// library for robust retry mechanisms with exponential backoff and jitter.
+package api
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+
+	"github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/errors"
+)
+
+// maxRetryAttempts defines the maximum number of retry attempts that will be made
+// for retryable HTTP responses (429 and 504). After this many failed attempts,
+// the last response will be returned to the caller without further retries.
+// Value: 5 attempts (6 total requests including the initial request).
+const maxRetryAttempts = 5
+
+// tflogBridge implements retryablehttp.Logger interface and forwards log messages to tflog.
+// This ensures consistent logging throughout the provider using Terraform's structured logging system.
+type tflogBridge struct {
+	ctx context.Context
+}
+
+// Printf implements the retryablehttp.Logger interface by forwarding to tflog.Debug.
+// This integrates retry logging with Terraform's logging system, allowing users to control
+// retry logging via TF_LOG environment variable alongside other provider logs.
+func (t *tflogBridge) Printf(format string, args ...interface{}) {
+	// Use tflog.Debug for retry messages as they are diagnostic information
+	// The [RETRY] prefix helps identify these logs among other provider logs
+	tflog.Debug(t.ctx, fmt.Sprintf("[RETRY] "+format, args...))
+}
+
+// customRetryPolicy implements retry logic for the Couchbase Capella API.
+// It handles specific cases for 429 (rate limiting) and 504 (gateway timeout) responses.
+//
+// Retry Behavior:
+//   - HTTP 429 (Too Many Requests): Always retries, respecting Retry-After header if present
+//   - HTTP 504 (Gateway Timeout): Retries unless the response body contains "code": 7001,
+//     which indicates a non-retryable error condition for index DDL operations
+//   - All other status codes: Returns immediately without retry
+//
+// For 504 responses with code 7001, it returns the specific ErrGatewayTimeoutForIndexDDL
+// error to provide better error context to callers, matching V1 client behavior.
+//
+//nolint:nilerr // returning nil error is expected behavior for retryablehttp.CheckRetry interface
+func customRetryPolicy(ctx context.Context, resp *http.Response, err error) (bool, error) {
+	// Handle connection errors - retry these
+	if err != nil {
+		return true, nil
+	}
+
+	switch resp.StatusCode {
+	case http.StatusTooManyRequests:
+		// Always retry 429 responses - retryablehttp handles Retry-After header automatically
+		return true, nil
+
+	case http.StatusGatewayTimeout:
+		// Check for special code 7001 (do not retry)
+		body, readErr := io.ReadAll(resp.Body)
```

Comment:
What if this is a large response? Do we want to read the full response body?

Reply:
should we just unmarshal here instead of reading the response?

Reply:
fixed in [c3348ea](https://github.com/couchbasecloud/terraform-provider-couchbase-capella/pull/404/commits/c3348ea650cd02a6ab17ff519377d38df1b5379a)